### PR TITLE
Support S32/U32 indices for BWD embedding & Neuron implicit downcast

### PR DESCRIPTION
**Pull request by rpsilva-aws**
In this PR, we extend embedding tensor operations to allow S32 indices. This follows suits with other operations, in order to add flexibility and potentially performance benefits for accelerator backends. Reference for embedding dense bwd: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/Embedding.cpp#L117

In addition, we also re-introduce the implicit downcasting for Neuron S64/U64 types, since the Neuron compiler does not support 64 bits.

There is an ongoing effort to further extend this requirement to other tensor operations involving indices: https://github.com/pytorch/pytorch/pull/142160. Once this is resolved, we adapt it on XLA as well.

### CODE REVIEW

File: torch_xla/csrc/dtype.cpp [hunk 1]
```diff
@@ -143,9 +143,11 @@ xla::PrimitiveType MaybeDowncastToXlaDeviceType(
       return CheckNeuronDevice(hw_type) ? xla::PrimitiveType::S32
                                         : xla::PrimitiveType::S16;
     case xla::PrimitiveType::S64:
-      return xla::PrimitiveType::S64;
+      return CheckNeuronDevice(hw_type) ? xla::PrimitiveType::S32
+                                        : xla::PrimitiveType::S64;
     case xla::PrimitiveType::U64:
-      return xla::PrimitiveType::U64;
+      return CheckNeuronDevice(hw_type) ? xla::PrimitiveType::U64
```
Comment (tengyifei) [inline torch_xla/csrc/dtype.cpp:149]: Is this check flipped?

**Review by tengyifei — CHANGES_REQUESTED** [review]
Is it possible to add a test at all?

File: torch_xla/csrc/dtype.cpp (hunk 1 above)
Comment (rpsilva-aws) [reply torch_xla/csrc/dtype.cpp:149]: Thanks!

**Review by tengyifei — APPROVED** [review]

**Comment by rpsilva-aws** [comment]
@tengyifei Ran yapf over the test file. PTAL, thanks!

**Review by tengyifei — APPROVED** [review]

**Pull request merged**